qutip/solver/integrator.py

```diff
@@ -0,0 +1,185 @@
+        self.system = system
+        self._stats = {}
+        self.options = options
+        self._is_set = False
+        self._prepare()
+
+    def _prepare(self):
+        """
+        Initialize the solver
+        """
+        raise NotImplementedError
+
+    def set_state(self, t, state0):
+        """
+        Set the state of the ODE solver.
+
+        Parameters
+        ----------
+        t : float
+            Initial time
+
+        state0 : qutip.Data
+            Initial state.
+        """
+        raise NotImplementedError
+
+    def integrate(self, t, step=False, copy=True):
+        """
+        Evolve to t.
+
+        If `step` advance up to t by one internal solver step.
+        Should be able to retreive the state at any time between the present
+        time and the resulting time with subsequent calls.
+
+        Before calling `integrate` for the first time, the initial step should
+        be set with `set_state`.
+
+        Parameters
+        ----------
+        t : float
+            Time to integrate to, should be larger than the previous time. If
+            the last integrate call was use with ``step=True``, the time can be
+            between the time at the start of the last call and now.
+
+        step : bool [False]
+            When True, integrate for a one internal step of the ODE solver.
+
+        copy : bool [True]
+            Whether to return a copy of the state or the state itself.
+
+        Return
+        ------
+        (t, state) : (float, qutip.Data)
+            The state of the solver at ``t``. The returned this can differ from
+            the input time only when ``step=True``.
+        """
+        raise NotImplementedError
+
+    def get_state(self, copy=True):
+        """
+        Obtain the state of the solver as a pair (t, state).
+
+        Return
+        ------
+        (t, state) : (float, qutip.Data)
+            The state of the solver at ``t``.
+        """
+        raise NotImplementedError
+
+    def run(self, tlist):
+        """
+        Integrate the system yielding the state for each times in tlist.
+
+        Parameters
+        ----------
+        tlist : *list* / *array*
+            List of times to yield the state.
+
+        Yield
+        -----
+        (t, state) : (float, qutip.Data)
+            The state of the solver at each ``t`` of tlist.
+        """
+        for t in tlist[1:]:
+            yield self.integrate(t, False)
+
+    def reset(self):
+        """Reset internal state of the ODE solver."""
+        if self._is_set:
+            self.set_state(*self.get_state())
+
+    def update_args(self, args):
+        """
+        Change the argument of the system.
+        Reset the ODE solver to ensure numerical validity.
+
+        Parameters
+        ----------
+        args : dict
+            New arguments
+        """
+        self.system.arguments(args)
+        self.reset()
+
+
+# TODO: These integrator set will be part of the solver classes
```

Comment:
Agreed that the add_integrators and these sets belongs closer to the solvers, but happy with this as an intermediate step.